### PR TITLE
COPS-862

### PR DESCRIPTION
**Pull request by marty331**
Implement new payload for processing upgrades
The basket payload should include the following information, per this API spec:

    event_id - The Stripe Event ID
    event_type - customer.subscription.upgrade OR customer.subscription.downgrade
    close_date - Date that the event occurs (when the customer makes the change to their subscription)
    customer_id - The Stripe Customer ID
    nickname_old - This is created from the old Product name and the old Plan interval
    nickname_new - This is created from the new Product name and the new Product interval
    plan_amount_old - This is the price of the old Plan
    plan_amount_new - This is the price of the new Plan
    proration_amount - Use the upcoming invoice API to retrieve the prorated amount to provide. It will be calculated by summing together the line item amounts found within the Invoice object https://stripe.com/docs/api/invoices/object
    interval - This is the interval for the new Plan
    current_period_end - Billing cycle end date; this will populate the {NextInvoiceDate}

    in the email body for reactivated subscriptions
    subscription_id - Subscription id - so we can group charges for a specific subscription
    invoice_number - Invoice Number (customer facing value, to be included on the customer's invoice sent via email)
    invoice_id - Unique identifier of the invoice from Stripe

### CODE REVIEW

File: src/hub/vendor/customer.py [hunk 1]
```diff
@@ -558,3 +590,93 @@ def get_reactivation_data(self) -> Dict[str, Any]:
...
+        )
+        if new_product_set_order > previous_product_set_order:
+            return "customer.subscription.upgrade"
+        elif previous_product_set_order > new_product_set_order:
+            return "customer.subscription.downgrade"
+        else:
+            return None
+
+    def get_previous_plan_amount(self, previous_product: Dict[str, Any]) -> int:
+        """
```
[the remark this answers is not in the document]
Comment (marty331) [reply src/hub/vendor/customer.py]: OK that makes sense...and done.

File: src/hub/vendor/customer.py [hunk 2]
```diff
@@ -558,3 +590,93 @@ def get_reactivation_data(self) -> Dict[str, Any]:
...
+            "get subscription type",
+            new_product_set_order=new_product_set_order,
+            previous_product_set_order=previous_product_set_order,
+        )
+        if new_product_set_order > previous_product_set_order:
+            return "customer.subscription.upgrade"
+        elif previous_product_set_order > new_product_set_order:
+            return "customer.subscription.downgrade"
+        else:
+            return None
```
[the remark this answers is not in the document]
Comment (marty331) [reply src/hub/vendor/customer.py]: I went back and forth on the placement of this.  Theoretically this shouldn't happen but you know it will.  OK, I moved it.

File: src/hub/vendor/customer.py [hunk 3]
```diff
@@ -394,25 +394,48 @@ def run(self) -> bool:
...
             logger.info("customer subscription new recurring", data=data)
             routes = [StaticRoutes.SALESFORCE_ROUTE]
         elif not current_cancel_at_period_end and previous_cancel_at_period_end:
-            data = self.create_payload("customer.subscription.reactivated", user_id)
+            data = self.create_payload(
+                "customer.subscription.reactivated", user_id, previous_plan=None
+            )
             logger.info("customer subscription reactivated", data=data)
             routes = [StaticRoutes.SALESFORCE_ROUTE]
+        elif previous_plan:
```
[the remark this answers is not in the document]
Comment (marty331) [reply src/hub/vendor/customer.py]: that makes sense, done.

File: src/hub/vendor/customer.py [hunk 4]
```diff
@@ -558,3 +590,93 @@ def get_reactivation_data(self) -> Dict[str, Any]:
...
+        logger.debug(
+            "sub change",
+            event_type=event_type,
+            payload=payload,
+            self_payload=self.payload,
+            upcoming_invoice=upcoming_invoice,
+            amount_due=upcoming_invoice.get("amount_due", 0),
+        )
+        return dict(
+            nickname_old=previous_plan.get("nickname", "Not available"),
```
[the remark this answers is not in the document]
Comment (marty331) [reply src/hub/vendor/customer.py]: You are right, I actually had it on my list to go back and do this. Done.

File: tox.ini [hunk 5]
```diff
@@ -45,8 +45,8 @@ commands =
     ; coverage report.
     ;
     ; `--no-cov-on-fail` provide no code coverage on a failing test run.
-    py.test --cov-config={toxinidir}/.coveragerc --cov-append --cov-branch --no-cov-on-fail --cov=src/sub -k src/sub --ignore=src/sub/shared
-    py.test --cov-config={toxinidir}/.coveragerc --cov-append --cov-branch --no-cov-on-fail --cov=src/hub -k src/hub --ignore=src/hub/shared
-    py.test --cov-config={toxinidir}/.coveragerc --cov-append --cov-branch --no-cov-on-fail --cov=src/shared -k src/shared
+    py.test --cov-report term-missing --cov-config={toxinidir}/.coveragerc --cov-append --cov-branch --no-cov-on-fail --cov=src/sub -k src/sub --ignore=src/sub/shared
```
Comment (bowlofstew) [inline tox.ini]: Given that we have been doing discrete pull requests (PR), should we extract this change into it's own PR that is separate from this workstream?

Comment (marty331) [reply tox.ini]: Good point. Done.

File: src/hub/tests/unit/stripe/customer/test_stripe_customer.py [hunk 6]
```diff
@@ -647,26 +651,49 @@ def setUp(self) -> None:
...
+            "src/hub/tests/unit/fixtures/stripe_sub_updated_event_change.json"
+        ) as fh:
+            self.subscription_change_event = json.loads(fh.read())
+
         with open(
             "src/hub/tests/unit/fixtures/stripe_sub_updated_event_no_trigger.json"
         ) as fh:
             self.subscription_updated_event_no_match = json.loads(fh.read())
 
+        with open("src/hub/tests/unit/fixtures/stripe_previous_plan1.json") as fh:
```
Comment (sidler-mozilla) [inline src/hub/tests/unit/stripe/customer/test_stripe_customer.py]: save the dir as `src/hub/tests/unit/fixtures` instead of copy and pasting it over and over a thousand times; this will also make the `with open` expressions shorter and not have `black` do the ugly wrap

File: src/hub/tests/unit/stripe/customer/test_stripe_customer.py [hunk 7]
```diff
@@ -647,26 +651,49 @@ def setUp(self) -> None:
...
+            self.previous_plan = json.loads(fh.read())
+
+        with open("src/hub/tests/unit/fixtures/valid_plan_response.json") as fh:
```
Comment (sidler-mozilla) [inline src/hub/tests/unit/stripe/customer/test_stripe_customer.py]: over and over

File: src/hub/tests/unit/stripe/customer/test_stripe_customer.py (hunk 6 above)
Comment (marty331) [reply src/hub/tests/unit/stripe/customer/test_stripe_customer.py]: done

File: src/hub/tests/unit/stripe/customer/test_stripe_customer.py (hunk 7 above)
Comment (marty331) [reply src/hub/tests/unit/stripe/customer/test_stripe_customer.py]: done and done and done
